feat: LEAP-1367: Add TimelineLabels to mark parts of the video

hlomzik opened the pull request:
- new control tag TimelineLabels with corresponding result type
- new TimelineRegion to store values in `ranges` field
- new data format, shorter than `sequence` to represent time spans
- adjustments to Video tag to support new kind of results
- UX to draw timeline regions on a special new line and on empty spaces in timeline
- fix deserialization to properly read new results
- improvements to Video UX:
  - region indices are now in sync with indices in regions list
  - hover and selected states are distinguisable now
  - timeline is not weirdly jumping when you selected a region
  - new `Video timelineHeight` param to adjust the timeline area

### PR fulfills these requirements
- [ ] Tests for the changes have been added/updated (for bug fixes/features)
- [ ] Docs have been added/updated (for bug fixes/features)
- [x] Best efforts were made to ensure docs/code are concise and coherent (checked for spelling/grammatical errors, commented out code, debug logs etc.)
- [x] Self-reviewed and ran all changes on a local instance (for bug fixes/features)

### Describe the reason for change
This is a helpful feature, requested for a long time. This first iteration is focused on instant and short time spans with not so many expected regions for annotation.

#### What alternative approaches were there?
per-frame choices, which are less flexible and have a worse fit into our architecture.

#### What feature flags were used to cover this change?
none, this feature considered an add-on, not changing much any core and existing functionality.

### Does this PR introduce a breaking change?
- [ ] Yes, and covered entirely by feature flag(s)
- [ ] Yes, and covered partially by feature flag(s)
- [x] No
- [ ] Not sure (briefly explain the situation below)

### What level of testing was included in the change?
tests will follow
- [ ] e2e
- [ ] integration
- [ ] unit

<!-- ELLIPSIS_HIDDEN -->

----

| :rocket: | This description was created by [Ellipsis](https://www.ellipsis.dev) for commit f1272b6838ed1f2e21f71ae447a91d8f627fc864  | 
|--------|--------|

### Summary:
Introduced `TimelineLabels` for video annotation, enabling frame classification and timeline region marking, with updates across models, components, and documentation.

**Key points**:
- Added `TimelineLabels` tag for video frame classification in `docs/source/tags/timelinelabels.md`.
- Introduced `TimelineRegion` model in `web/libs/editor/src/regions/TimelineRegion.js` to store time ranges.
- Updated `Video` tag in `web/libs/editor/src/tags/object/Video/Video.js` to support `timelineHeight` parameter.
- Enhanced `Timeline` component in `web/libs/editor/src/components/Timeline/Timeline.tsx` to handle new drawing events.
- Modified `Frames` view in `web/libs/editor/src/components/Timeline/Views/Frames/Frames.tsx` to support timeline region drawing.
- Updated `Result` model in `web/libs/editor/src/regions/Result.js` to include `timelinelabels`.
- Adjusted `Label` component in `web/libs/editor/src/tags/control/Label.jsx` to support `TimelineLabels`.
- Added tests for video timeline seek indicator in `web/libs/editor/tests/e2e/tests/regression-tests/video-timeline-seek-indicator.test.js`.


----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)


<!-- ELLIPSIS_HIDDEN -->

hlomzik committed:
feat: LEAP-1362: Add TimelineLabels to mark parts of the video

- new control tag TimelineLabels with corresponding result type
- new TimelineRegion to store values in `ranges` field
- new data format, shorter than `sequence` to represent timespans
- adjustments to Video tag to support new kind of results
- UX draft to draw timeline regions at any point, will definitely be improved
- fix deserialization to properly read new results

hlomzik committed:
Add special appearance for timeline regions

Now they are timespans with distinct borders

hlomzik committed:
Properly serialize timelinelabels

Video tag has `mergeLabelsAndResults: true` so it created extra `labels` property, messing with results.
Now there are 3 cases:
- usual tags and regions: all results
- video with rectangles: labels added to rectangles
- video with timeline labels: usual labels result with extra values

hlomzik committed:
Tiny fixes

- fix types
- remove excess hooks
- react on video region changes

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/10361611758

hlomzik committed:
Allow to draw only on empty space or special line

hlomzik committed:
Add quazi-line to create new regions

Add it when the label is selected but no regions or selected,
so user has intent to create a new region with label.

hlomzik committed:
Merge remote-tracking branch 'origin/fb-leap-1362/timeline-regions' into fb-leap-1362/timeline-regions

hlomzik committed:
Fix drawing on empty space (allowed) and region label (forbidden)

hlomzik committed:
Show correct region index like in Outliner

hlomzik committed:
Trigger CI

hlomzik commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/10421772301)
> Successfully merged:  create mode 100644 web/libs/frontend-test/src/helpers/LSF/RichText.ts

hlomzik committed:
Merge branch 'develop' into 'fb-leap-1362/timeline-regions'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/10421772301

hlomzik committed:
Add `timelineHeight` param to Video tag

hlomzik committed:
Fix missing index passing + some optional types

hlomzik committed:
Add only the final state to undo history after drawing the region

hlomzik committed:
Don't break existing Video rectangles regions

Split controls, detect correct video region type, etc.

hlomzik committed:
Suppress timeline labels behaviour if there are no TimelineLabels

New region was created every time user worked with video rectangle on timeline.
That was because timeline behaviour kicked in even without proper control in place.

hlomzik commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/10477666123)
> Successfully merged:  delete mode 100644 web/libs/datamanager/src/components/Common/TableOld/utils.js

hlomzik committed:
Merge branch 'develop' into 'fb-leap-1362/timeline-regions'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/10477666123

hlomzik committed:
Fix conditions for drawing frames in the timeline

Much better check for not being on labels area.
Plus no excess check for classname, it's already checked by `!regionRow`

hlomzik committed:
Fix labels behaviour

Create regions only when labels are selected.

hlomzik committed:
Make TimelineLabels to behave like regular labels

Apply labels in a regular way via `resultValue` in `createResult()`.
Respect “select region after” and “keep label selected” settings.
Add comments about multi-labels.

hlomzik committed:
Disable extra controls for timeline regions

Toggle Keypoint and Toggle Interpolation are disabled for them.

hlomzik committed:
Reverse regions only if TimelineLabels is presented

To not break the usual flow with video rectangles.

hlomzik committed:
Fix and improve docs about TimelineLabels

caitlinwheeless committed:
Added a screenshot and a bit of text

caitlinwheeless committed:
Fixing image link

caitlinwheeless committed:
Making fixes per Andrew's feedback

hlomzik committed:
Fix RichText docs to unblock docs auto-generation

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/10532047661

caitlinwheeless committed:
Update wording for results JSON

bmartel commented on web/libs/editor/src/regions/TimelineRegion.js:
```diff
@@ -0,0 +1,109 @@
+    setSequence(sequence) {
+      const [start, end] = sequence;
+
+      self.parent.annotation.history.setReplaceNextUndoState();
```
Maybe adding a comment here about why this is needed before setting self.ranges. I know why you have to do it, but will likely forget in the future 😆. Every time I see this I barely recall that the changes made to the data properties will cause the TimeTravel (undo/redo) recorder to capture it as a change unless you opt out of it.
```
      self.parent.annotation.history.setReplaceNextUndoState();
```

bmartel reviewed (approved):
Well done, and well commented! There was only one piece which I called out for maybe another comment to describe it's need.

bmartel commented on web/libs/editor/src/components/Timeline/Views/Frames/Keypoints.styl:
Just a reminder that all the stylus files are now converted, and when you merge develop into this branch to carry forward your changes into the Keypoints.scss file.

hlomzik replied:
renamed it, thank you!

Gondragos commented on web/libs/editor/src/components/Timeline/Types.ts:
```diff
@@ -73,17 +77,22 @@ export interface TimelineViewProps {
   selected: boolean;
   sequence: TimelineRegionKeyframe[];
+  /** is this timeline region with spans */
+  timeline?: boolean;
```
Shouldn't it be something like `isTimeline`?

hlomzik replied:
I'm following the style of other props here (`selected`, `visible`) and I think that for small internal types it's okay

Gondragos commented on web/libs/editor/src/regions/Area.js:
```diff
@@ -57,12 +60,19 @@ const Area = types.union(
 
+      // @todo dirty hack to distinguish two video types
+      if (tag.type === "video") {
+        if (sn.sequence) return VideoRectangleRegionModel;
```
It looks like this condition is fine for drawing a new region but will not work with restoring regions from the previously created annotation.

hlomzik replied:
yeah, fixed this, thank you! still feels weird, but current way of detecting region type is definitely not the ideal

Gondragos commented on web/libs/editor/src/tags/object/Video/Video.js:
```diff
@@ -198,28 +201,46 @@ const Model = types
 
       addRegion(data) {
-        const control = self.videoControl() ?? self.control();
+        const control = self.videoControl ?? self.timelineControl;
```
If we have more than one control in the config we do not get the correct one here.

hlomzik replied:
Changed control getters to be explicit and split their usage — now it's `addVideoRegion()` and `addTimelineRegion()`

Gondragos commented on web/libs/editor/src/components/Timeline/Views/Frames/Frames.tsx:
```diff
@@ -204,22 +212,34 @@ export const Frames: FC<TimelineViewProps> = ({
+
+        if (region) {
+          const [start, end] = frame > baseFrame ? [baseFrame, frame] : [frame, baseFrame];
+          region.setSequence([{ frame: start, enabled: true }, { frame: end, enabled: false }]);
```
It could be not a `TimelineRegion` so `setSequence` could not exist. (At least it is so when we have mixed config, like `VideoRectangle` + `TimelineLabels`)

hlomzik replied:
thank you! the region here could be only timeline region now, I fixed drawing method + I renamed method to `setRanges()`

Gondragos reviewed (changes requested):
We at least need to fix this condition, or we will not be able to work with `VideoReactangle`.
```
if (sn.sequence) return VideoRectangleRegionModel;
```
https://github.com/HumanSignal/label-studio/pull/6191/files#diff-c27478dffecaefcefc82458645209521f9e2983a6f53c3726178630a69d2aa10R64-R67

hlomzik committed:
Small comment about undo history management

hlomzik committed:
Split adding rectangle and timeline regions

hlomzik committed:
Merge branch 'develop' into fb-leap-1362/timeline-regions

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/10598173923

hlomzik commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/10607022798)
> Successfully merged: Already up to date.

hlomzik committed:
Fix issues with srubbing and hopping in Video

- fix usage of wrong units:
  - positions are 1-based
  - offsets are 0-based, so in calculations positions should be deducted by 1
  - positions are offsets * steps, they were mixed up together
- calculate start and end of currently visible window correctly
  to hop back and forth without skipping frames
- fix conditions for hopping, hop back was unreachable
- fix scrubbing on random page when it's not aligned with `framesInView`;
  jump to next page only if we are moving outside of current view
- fix value stored as previous scroll to be the correct unit
- don't reset current position when video is not fully loaded
- add more meaningful comments

hlomzik commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/10685592036)
> Successfully merged:  create mode 100644 .github/workflows/validator-poetry-version.yml

hlomzik committed:
Merge branch 'develop' into 'fb-leap-1362/timeline-regions'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/10685592036

hlomzik committed:
Tiny fixes, comments, renamed setSequence -> setRanges

hlomzik committed:
Fix region restoration for VideoRectangles

Fix region type detection in our crazy dispatcher in Area.

hlomzik committed:
Fix test to move indicator window back and forth

We don't focus on half-visible frames now, so that had to be fixed.
Plus now test logic is more flexible and bulletproof.

hlomzik committed:
Fix hovered and selected state for all regions in video timeline

hlomzik committed:
Merge branch 'develop' into fb-leap-1362/timeline-regions

MihajloHoma commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/10734607181)
> Successfully merged: Already up to date.

MihajloHoma commented:
/docker build
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/10736594322)
> Docker image was pushed with the tag `20240906.101645-fb-leap-1362-timeline-reg-ab969e8cc`

hlomzik committed:
Fix linting

hlomzik committed:
Style lint fixes

robot-ci-heartex committed:
ci: Build frontend

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/10773751099

hlomzik commented:
@ellipsis-dev review this
